doc/feature/compartmentReport.md

```diff
@@ -0,0 +1,422 @@
+         * when read is called the result is undefined.
+         */
+        boost::future< floatsPtr > readNextFrame( );
+
+        /**
+         * Forward to a given absolute timestamp.
+         *
+         * Preconditions:
+         * - r.getState() is OK or ENDED.
+         * - The report was open in read mode.
+         * - There is no previous read or seek operation with a pending future.
+         * - toTimestamp >= getStartTime()
+         *
+         * Postconditions:
+         * Let:
+         *  - r be the CompartmentReport
+         *  - f be the returned future by r.seek(toTimestamp)
+         *  After f.wait() returns:
+         *  - r.getCurrentTime() ~= toTimestamp rounded to the nearest multiple
+              of the timestep, ...
+         *  - ... getState() == OK -> r.getEndTime() > getCurrentTime() and ...
+         *  - ... getState() == ENDED -> r.getEndTime() == getCurrentTime()
+         *
+         * After successful f.wait_for or f.wait_until, the result is the same
+         * as above. If they time out, no observable state changes.
+         *
+         * @throw std::runtime_error if a precondition does not hold.
+         */
+        boost::future< void > seek( const float toTimestamp );
+
+        /** Write the header information of this report.
+         *
+         * Preconditions:
+         * - The report was open in write mode.
+         * - The header has not already been written.
+         *
+         * @param startTime the start time of the report
+         * @param endTime the end time of the report
+         * @param timestep the timestep between report frames
+         * @param dunit the unit of the data, e.g. mV
+         * @param tunit the unit of the time, e.g. ms
+         * @throw std::invalid_argument if any passed argument is invalid
+         */
+        void writeHeader( const float startTime, const float endTime,
+                          const float timestep, const std::string& dunit,
+                          const std::string& tunit );
+
+        /** Write the compartment counts for each section for one cell.
+         *
+         * Preconditions:
+         * - The report was open in write mode.
+         * - writeHeader has already been called.
+         *
+         * @param gid the GID of the cell
+         * @param counts the number of compartments per section
+         * @throw std::runtime_error if writing was not successful.
+         */
+        void writeMapping( const uint32_t gid,
+                           const uint16_ts& counts );
+
+        /** Write the compartment counts for each section for one cell.
+         *
+         * Preconditions:
+         * - The report was open in write mode.
+         * - writeMapping has already been called.
+         *
+         * @param gid the GID of the cell
+         * @param voltages the voltages per compartment to save
+         * @throw std::runtime_error if writing was not successful.
```

Comment:
Mmm... I can't think of a specific example now. But in general, my idea is that exceptions are for really exceptional behaviors, as their impact is usually more critical than a return value, and their handling a bit more complicated.
In the case of read/write functions, as you said, they can either succeed or fail, and those are very valid results that usually happen often.

There's no reason to use both exceptions and return values, true (maybe only if, internally, there is something specific that exceptionally failed and you want to make sure that is handled, then the exception can be raised). But in my opinion, and without knowing about the best practices in this context, I guess I tend to prefer return values. That said, I have no strong preference, and since you have much more experience in the usage of this kind of apis, you are in a better position to decide :)